File: src/main/java/com/aquasecurity/plugins/trivy/ui/treenodes/FileTreeNode.java
package com.aquasecurity.plugins.trivy.ui.treenodes;

import com.aquasecurity.plugins.trivy.model.Result;

import javax.swing.*;
import javax.swing.tree.DefaultMutableTreeNode;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.Optional;

public class FileTreeNode extends DefaultMutableTreeNode implements TrivyTreeNode {

    private String target;
    private String type;
    private String className;

    public FileTreeNode(Result result) {
        if (result == null) {
            return;
        }
        this.target = result.target;
        this.type = result.type;
        this.className = result.ClassName;

        update(result);
    }

public void update(Result result) {

        if (result.misconfigurations != null && result.misconfigurations.size() > 0) {
            List<MisconfigurationTreeNode> visited = new ArrayList<>();
            result.misconfigurations.forEach(ms -> {

                if (visited.stream().noneMatch(v -> Objects.equals(v.misconfiguration.id, ms.id))) {
                    visited.add(new MisconfigurationTreeNode(ms));
                }
                Optional<MisconfigurationTreeNode> findingNode = visited.stream().filter(vf -> Objects.equals(vf.misconfiguration.id, ms.id)).findFirst();
                if (findingNode.isPresent()) {
                    MisconfigurationTreeNode node = findingNode.get();
                    node.add(new LocationTreeNode(result.target, result.type, ms));
                }
            });

            visited.forEach(this::add);
        }

        if (result.vulnerabilities != null && result.vulnerabilities.size() > 0) {
            List<VulnerablePackageTreeNode> visited = new ArrayList<>();
            result.vulnerabilities.forEach(vulnerability -> {

                if (visited.stream().noneMatch(v -> Objects.equals(v.vulnerability.pkgName, vulnerability.pkgName))) {
                    visited.add(new VulnerablePackageTreeNode(vulnerability));
                }
                Optional<VulnerablePackageTreeNode> findingNode = visited.stream().filter(vf -> Objects.equals(vf.vulnerability.pkgName, vulnerability.pkgName)).findFirst();
                if (findingNode.isPresent()) {
                    VulnerablePackageTreeNode node = findingNode.get();
                    node.add(new VulnerabilityTreeNode(vulnerability, result.target));
                }
            });

            visited.forEach(this::add);
        }

        if (result.secrets != null && result.secrets.size() > 0) {
            List<SecretTreeNode> visited = new ArrayList<>();
            result.secrets.forEach(secret -> {
                if (visited.stream().noneMatch(v -> Objects.equals(v.secret.ruleID, secret.ruleID))) {
                    visited.add(new SecretTreeNode(secret, result.target));
                }
                Optional<SecretTreeNode> findingNode = visited.stream().filter(vf -> Objects.equals(vf.secret.ruleID, secret.ruleID)).findFirst();
                if (findingNode.isPresent()) {
                    SecretTreeNode node = findingNode.get();
                    //node.add(new SecretTreeNode(secret, result.target));
                }
            });

            visited.forEach(this::add);
        }


    }

    @Override
    public Icon getIcon() {
        String iconType = this.type;
        if (iconType == null || Objects.equals(iconType, "")) {
            iconType = this.className;
        }
        return IconHelper.getFileIcon(iconType);
    }

    @Override
    public String getTitle() {
        return this.target;
    }

    @Override
    public String getTooltip() {
        return "";
    }

    public String getTarget() {
        return this.target;
    }

}

File: src/main/kotlin/com/aquasecurity/plugins/trivy/ui/treenodes/FileTreeNode.kt
package com.aquasecurity.plugins.trivy.ui.treenodes

import com.aquasecurity.plugins.trivy.model.report.Misconfiguration
import com.aquasecurity.plugins.trivy.model.report.Result
import com.aquasecurity.plugins.trivy.model.report.Secret
import com.aquasecurity.plugins.trivy.model.report.Vulnerability
import java.nio.file.Paths
import java.util.function.Consumer
import javax.swing.Icon
import javax.swing.tree.DefaultMutableTreeNode

class FileTreeNode(result: Result?) : DefaultMutableTreeNode(), TrivyTreeNode {
  var target: String = result!!.target!!
  private val type: String? = result!!.type
  private val className: String? = result!!.ClassName

  private val project =
    com.intellij.openapi.project.ProjectManager.getInstance().openProjects.firstOrNull()

  init {
    update(result!!)
  }

  fun update(result: Result) {
    if (result.misconfigurations != null && result.misconfigurations!!.isNotEmpty()) {
      val visited: MutableList<LocationTreeNode> = ArrayList()
      // sort the misconfigurations
      result.misconfigurations = result.misconfigurations!!.sortedBy { it.severity }
      result.misconfigurations!!.forEach(
        Consumer { ms: Misconfiguration ->
          // get count of misconfigurations with the same title
          val count =
            result.misconfigurations!!
              .stream()
              .filter { v: Misconfiguration -> v.title == ms.title }
              .count()
          if (visited.stream().noneMatch { v: LocationTreeNode -> v.title == ms.title }) {
            visited.add(LocationTreeNode(result.target.toString(), result.type.toString(), ms))
          }
          if (count > 1) {
            val findingNode =
              visited.stream().filter { vf: TrivyTreeNode -> vf.title == ms.title }.findFirst()
            if (findingNode.isPresent) {
              val node = findingNode.get()
              node.add(LocationTreeNode(result.target.toString(), result.type.toString(), ms, true))
            }
          }
        }
      )
      visited.forEach(Consumer { newChild: LocationTreeNode? -> this.add(newChild) })
    }

    if (result.vulnerabilities != null && result.vulnerabilities!!.isNotEmpty()) {
      val visited: MutableList<TrivyTreeNodeImpl> = ArrayList()
      // sort the vulnerabilities
      result.vulnerabilities = result.vulnerabilities!!.sortedBy { it.severity }
      result.vulnerabilities!!.forEach(
        Consumer { vulnerability: Vulnerability ->
          if (
            visited.stream().noneMatch { v: TrivyTreeNodeImpl -> v.title == vulnerability.pkgName }
          ) {
            visited.add(TrivyTreeNodeImpl(result.type.toString(), vulnerability))
          }
          val findingNode =
            visited
              .stream()
              .filter { vf: TrivyTreeNodeImpl -> vf.title == vulnerability.pkgName }
              .findFirst()
          if (findingNode.isPresent) {
            val node = findingNode.get()
            node.add(
              LocationTreeNode(result.target.toString(), result.type.toString(), vulnerability)
            )
          }
        }
      )
      visited.forEach(Consumer { newChild: TrivyTreeNodeImpl? -> this.add(newChild) })
    }

    if (result.secrets != null && result.secrets!!.isNotEmpty()) {
      val visited: MutableList<LocationTreeNode> = ArrayList()
      // sort the secrets
      result.secrets = result.secrets?.sortedBy { it.title }
      result.secrets?.forEach(
        Consumer { secret: Secret ->
          if (visited.stream().noneMatch { v: LocationTreeNode -> v.title == secret.title }) {
            visited.add(LocationTreeNode(result.target.toString(), result.type.toString(), secret))
          }
        }
      )
      visited.forEach(Consumer { newChild: LocationTreeNode? -> this.add(newChild) })
    }

    if (result.sasts != null && result.sasts!!.isNotEmpty()) {
      val visited: MutableList<LocationTreeNode> = ArrayList()
      // sort the SAST findings
      result.sasts = result.sasts!!.sortedBy { it.title }
      result.sasts!!.forEach(
        Consumer { sast ->
          val title = "${sast.title}:${sast.startLine}"
          if (visited.stream().noneMatch { v: LocationTreeNode -> v.title == title }) {
            val projectRoot = project!!.basePath // or project.getBasePath()
            val targetPath = result.target.toString()
            val relativePath =
              if (projectRoot != null) {
                Paths.get(projectRoot).relativize(Paths.get(targetPath)).toString()
              } else {
                targetPath // fallback if project root is not available
              }
            this.target = relativePath
            visited.add(LocationTreeNode(relativePath, result.type.toString(), sast))
          }
        }
      )
      visited.forEach(Consumer { newChild: LocationTreeNode? -> this.add(newChild) })
    }
  }

  override val icon: Icon?
    get() {
      var iconType = this.type
      if (iconType == null || iconType == "") {
        // get the extension from the file
        val extension = this.target.substring(this.target.lastIndexOf(".") + 1)
        if (extension != "") {
          iconType = extension
        } else {
          iconType = this.className
        }
      }
      return IconHelper.getFileIcon(iconType.toString())
    }

  override val title: String?
    get() = this.target

  override val tooltip: String?
    get() = ""
}
